src/backend/distributed/metadata/dependency.c

```diff
@@ -1097,11 +1097,33 @@ GetDependingViews(Oid relationId)
 	List *dependingViews = NIL;
 	List *nodeQueue = list_make1(tableNode);
 	ViewDependencyNode *node = NULL;
+	ViewDependencyNode *dependingNode = NULL;
+
+#if PG_VERSION_NUM >= PG_VERSION_13
+
+	/*
+	 * We do not use foreach here because we may insert new nodes to the list,
+	 * and there is a risk of having the list repalloc'ed in PG13.
+	 *
+	 * For more information, see postgres commit with sha
+	 * 1cff1b95ab6ddae32faa3efe0d95a820dbfdc164
+	 * that changed the representation of Lists to expansible arrays, not chains
+	 * of cons-cells.
+	 */
+	for (int dependingNodesPosition = 0; dependingNodesPosition < list_length(nodeQueue);
+		 dependingNodesPosition++)
+	{
+		node = list_nth(nodeQueue, dependingNodesPosition);
+		for (int dependingNodePosition = 0; dependingNodePosition < list_length(
+				 node->dependingNodes); dependingNodePosition++)
+		{
+			dependingNode = list_nth(node->dependingNodes, dependingNodePosition);
+#else
 	foreach_ptr(node, nodeQueue)
 	{
-		ViewDependencyNode *dependingNode = NULL;
 		foreach_ptr(dependingNode, node->dependingNodes)
 		{
+#endif
```

Comment:
why keep the old implementation?

Reply:
I think new implementation will cost O(n^2) unnecessarily for pg versions < 13 due to list implementation.

Reply:
> I think new implementation will cost O(n^2) unnecessarily for pg versions < 13 due to list implementation.


can you clarify this a bit  more? We do o(n^2) for both cases?

Reply:
`list_nth()` implementation in PG12 required O(N) operations to iterate through the chain of linked list elements (https://github.com/postgres/postgres/blob/97c39498e5ca9208d3de5a443a2282923619bf91/src/backend/nodes/list.c#L399-L400)

whereas on PG13 it is as simple as an O(1) array access (https://github.com/postgres/postgres/blob/67b2ceea01576933a1dc881ef6a65403e03483ee/src/include/nodes/pg_list.h#L269-L270)

I can possibly update the `foreach_ptr` macro to use the best methods for different PG versions to encapsulate this logic.

Reply:
That makes sense, though this code should have used dlist_foreach_modify.